### DISCUSSION TITLE
The furture of planners

### DISCUSSION THREAD

**Discussion by dotned** [microsoft/semantic-kernel#6981] [Q&A] [answered]
Can it be clarified what the future of planners is. I have seen a couple of recent messages discussing planners #6860 and #6735, with the first explicitly stating...

"Hello, we will move the Handlebars template out of experimental to non-experimental - we recommend you continue to use this."

However I have been readoing this page (https://learn.microsoft.com/en-us/semantic-kernel/concepts/planning?pivots=programming-language-csharp) updated only two days ago that states...

"Both the Stepwise and Handlebars planners will be deprecated in a future release of Semantic Kernel."

"If you are building a new AI agent, we recommend that you not use the Stepwise or Handlebars planners. Instead, use function calling as it is more powerful and easier to use."

Could there be some clarity on the future of planners please?

**Comment by dmytrostruk** [comment, marked as answer]
Hi @dotned !
The message _"Hello, we will move the Handlebars template out of experimental to non-experimental - we recommend you continue to use this."_ is related to Handlebars template, not planner. You can use template rendering in SK to build powerful prompts, which can execute other functions, have loops, conditional statements etc. 

Handlebars planner is built using Handlebars syntax, but it's separate functionality from template rendering. As for planners, we are going to create an example how to use function calling functionality and Python code interpreter to achieve similar results that planners produce, but in more reliable way. Thanks!

**Reply by dotned** [reply]
Hi, absolutely right, my mistake I had been deep in reading about planners, Stepwise and Handlebars, and when reading that MS page in relation to the point on depricating planners had crossed my wires. I'm aware Handlebars is a template format, used in the planner.

I'll be intestered in the example, especially when considering one the of key features on the Handlebars planner was to allow saving of the plan (template) for for reuse, and thus avoiding the expense of having the LLM recompute each time.

Thanks for replying so quickly.

**Reply by dmytrostruk** [reply]
> I'll be intestered in the example, especially when considering one the of key features on the Handlebars planner was to allow saving of the plan (template) for for reuse, and thus avoiding the expense of having the LLM recompute each time.

@dotned Thanks for sharing this information! As far as I can see, with function calling you can serialize and store chat history with information which functions to call, in what order etc. It should be possible to deserialize it then and execute functions as specified. We will try to include that in example.

**Reply by charleswm** [reply]
FWIW, I also had the same reaction as dotned when I saw the note on handlebars _**planners**_ deprecation in the docs. It may be worthwhile to update the docs to explicitly note the distinction between handlebars planning and handlebars templates.

**Reply by tojuasp** [reply]
Does the function calling example regarding planners already exist? Currently trying to learn how to implement instructions for accomplishing specific goal with certain quality, some examples of idiomatic approaches with SK would be appreciated, with C#.

**Reply by dmytrostruk** [reply]
@tojuasp Yes, we have examples how to migrate from StepwisePlanner to function calling, so you can compare both approaches side-by-side:
- https://github.com/microsoft/semantic-kernel/tree/main/dotnet/samples/Demos/StepwisePlannerMigration
- https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/Planners/AutoFunctionCallingPlanning.cs

**Comment by lgmorand** [comment]
on the same topic, regarding "python planner", blog says 

> To support local development, we’ll provide an out-of-the-box container for local testing. For production deployments, we’ll provide a connection to Azure Container Apps dynamic sessions.


Will it be a (MS) managed ACA or will we need to deploy an ACA env everytime we plan to use this (normally) more efficient planner ? What about private exposed ACA, how can we reach them from an heavy app for instance ?
